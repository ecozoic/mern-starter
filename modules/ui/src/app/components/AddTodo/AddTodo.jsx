import React from 'react';
import PropTypes from 'prop-types';
import { Button, Input, Icon } from 'semantic-ui-react';

/**
 * Manages the add todo form
 */
class AddTodo extends React.Component {
  static propTypes = {
    onInit: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
  };

  /**
   * Create an Add Todo component
   * @param {Object} props - component props
   * @param {Function} props.onInit - function to call during initialization/mounting
   * @param {Function} props.onSubmit - function to call on form submission
   */
  constructor(props) {
    super(props);

    this.state = {
      todo: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    // TODO: only fetch if not initialized
    this.props.onInit();
  }

  handleSubmit(event) {
    const { onSubmit } = this.props;
    const { todo } = this.state;

    event.preventDefault();

    if (!todo || !todo.trim()) {
      return;
    }

    onSubmit(todo);

    this.setState({ todo: '' });
  }

  handleChange(event) {
    this.setState({ todo: event.target.value });
  }

  render() {
    const { todo } = this.state;

    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <Input
            name="todo"
            type="text"
            value={todo}
            onChange={this.handleChange}
            action={
              <Button
                type="submit"
                primary
                animated
                disabled={!todo}
              >
                <Button.Content visible>
                  Add Todo
                </Button.Content>
                <Button.Content hidden>
                  <Icon name="plus" />
                </Button.Content>
              </Button>
            }
          />
        </form>
      </div>
    );
  }
}

export default AddTodo;
