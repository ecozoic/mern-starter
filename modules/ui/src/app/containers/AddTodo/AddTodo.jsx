import { connect } from 'react-redux';

import { addTodo, fetchTodos } from '../../actions';
import AddTodo from '../../components/AddTodo';

const mapDispatchToProps = dispatch => ({
  onSubmit: (todo) => {
    dispatch(addTodo(todo));
  },
  onInit: () => {
    dispatch(fetchTodos());
  },
});

/**
 * React-redux container for Add Todo component
 */
const AddTodoContainer = connect(
  undefined,
  mapDispatchToProps,
)(AddTodo);

export default AddTodoContainer;
