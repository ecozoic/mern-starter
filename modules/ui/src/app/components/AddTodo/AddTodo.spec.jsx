import React from 'react';
import _ from 'lodash';
import { mount, shallow } from 'enzyme';

import AddTodo from './';

describe('<AddTodo />', () => {
  it('renders without crashing', () => {
    const addTodo = shallow(
      <AddTodo
        onInit={_.identity}
        onSubmit={_.identity}
      />,
    );

    expect(addTodo).toMatchSnapshot();
  });

  describe('handleSubmit()', () => {
    it('prevents default behavior on form submission', () => {
      const preventDefault = jest.fn();

      const event = {
        preventDefault,
      };

      const addTodo = shallow(
        <AddTodo
          onInit={_.identity}
          onSubmit={_.identity}
        />,
      );

      const form = addTodo.find('form');
      form.simulate('submit', event);

      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it('noops for empty and whitespace strings', () => {
      const onSubmit = jest.fn();

      const addTodo = shallow(
        <AddTodo
          onInit={_.identity}
          onSubmit={onSubmit}
        />,
      );

      addTodo.setState({ todo: '' });

      const instance = addTodo.instance();
      const event = { preventDefault: _.identity };

      instance.handleSubmit(event);

      expect(onSubmit).not.toHaveBeenCalled();

      addTodo.setState({ todo: '    ' });

      instance.handleSubmit(event);

      expect(onSubmit).not.toHaveBeenCalled();
    });

    it('calls onSubmit for valid todos', () => {
      const onSubmit = jest.fn();
      const todo = 'todo';

      const addTodo = shallow(
        <AddTodo
          onInit={_.identity}
          onSubmit={onSubmit}
        />,
      );

      addTodo.setState({ todo });

      const instance = addTodo.instance();
      const event = { preventDefault: _.identity };

      instance.handleSubmit(event);

      expect(onSubmit).toHaveBeenCalledTimes(1);
      expect(onSubmit).toHaveBeenCalledWith(todo);
    });

    it('resets state after submission', () => {
      const addTodo = shallow(
        <AddTodo
          onInit={_.identity}
          onSubmit={_.identity}
        />,
      );

      addTodo.setState({ todo: 'foo' });

      const instance = addTodo.instance();
      const event = { preventDefault: _.identity };

      instance.handleSubmit(event);

      expect(addTodo.state('todo')).toEqual('');
    });
  });

  describe('handleChange()', () => {
    it('sets state', () => {
      const addTodo = shallow(
        <AddTodo
          onInit={_.identity}
          onSubmit={_.identity}
        />,
      );

      const instance = addTodo.instance();
      const todo = 'todo';
      const event = { target: { value: todo } };

      instance.handleChange(event);

      expect(addTodo.state('todo')).toEqual(todo);
    });
  });

  describe('componentDidMount', () => {
    it('calls onInit after mounting', () => {
      const onInit = jest.fn();

      mount(
        <AddTodo
          onInit={onInit}
          onSubmit={_.identity}
        />,
      );

      expect(onInit).toHaveBeenCalledTimes(1);
    });
  });
});
