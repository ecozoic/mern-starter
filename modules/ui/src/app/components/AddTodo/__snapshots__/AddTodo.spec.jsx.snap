// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTodo /> renders without crashing 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <Input
      action={
        <Button
          animated={true}
          as="button"
          disabled={true}
          primary={true}
          type="submit"
        >
          <ButtonContent
            visible={true}
          >
            Add Todo
          </ButtonContent>
          <ButtonContent
            hidden={true}
          >
            <Icon
              as="i"
              name="plus"
            />
          </ButtonContent>
        </Button>
      }
      name="todo"
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
</div>
`;
